pkg/db/db.go

```diff
@@ -0,0 +1,104 @@
+package db
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+	"go.uber.org/zap"
+)
+
+func GetConnectionString(host, user, password, name string, port int) string {
+	return fmt.Sprintf(
+		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
+		host,
+		port,
+		user,
+		password,
+		name,
+	)
+}
+
+type Database struct {
+	tables struct {
+		providers string
+	}
+	pool *pgxpool.Pool
+}
+
+func NewClient(ctx context.Context, connString string, tables map[string]string) (*Database, error) {
+	pool, err := pgxpool.New(ctx, connString)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new instace of pool: %v", err)
+	}
+
+	if err := pool.Ping(ctx); err != nil {
+		return nil, fmt.Errorf("failed to ping database: %v", err)
+	}
+
+	if _, ok := tables["providers"]; !ok {
+		return nil, fmt.Errorf("incomplete table name mapping, missing a table name for \"providers\" table")
+	}
+
+	db := Database{}
+	db.pool = pool
+	db.tables.providers = tables["providers"]
+	return &db, nil
+}
+
+func (pg *Database) Close() {
+	pg.pool.Close()
+}
+
+func (pg *Database) OnProvider(ctx context.Context, logger *zap.Logger, notifyExisting bool, fn func(providerName string)) error {
+	if notifyExisting {
+		go func() {
+			query := fmt.Sprintf("SELECT * FROM %s", pg.tables.providers)
+			rows, err := pg.pool.Query(ctx, query)
+			if err != nil {
+				logger.Fatal("failed to query database", zap.Error(err))
+			}
+			defer rows.Close()
+
+			for rows.Next() {
+				var c types.ProviderCluster
+				if err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status); err != nil {
+					logger.Fatal("failed to scan row", zap.Error(err))
+				}
+				fn(c.ClusterID)
+			}
+		}()
+	}
```

Comment:
Running in a go routine can create a race condition between events based on existing records in the DB and new even coming from the listen. In theory, we have Forman taking care of parallel events on the same provider record, but I think we should not rely on other packages to solve problems that we can easily solve within this package. 

In addition, OnProvider will not return if it gets to the WaitForNotification part of the function (below). This means control will never go back to the caller func, which is not something I expect from a function that is supposed to set up a handler for an event/notification.  

I will suggestion to have the entire logic invoked in a single Go routine (from the start till the end) and not separate the first part into its own Go routine to make sure we preserve order.

@Madhu-1 WDYT?

Reply:
The reason for handling the already existing data in database and new events separately for below reason

* From my understanding, we scan all the entries from the database and process them serially we might miss below events
* If there is an update for the already processed provider
* If there is a new provider added

The above are the reason to start a listener and consider handling the objects in the database in a goroutine and as you mentioned depend on the forman to take care of for not for parallel processing a single provider.

Let me know what do you think, if we can discard above 2 cases I can have a single goroutine for this one.

Reply:
Just to understand, you concern is that between the time we are reading the records from the DB and the time we start listening to notifications there might be a new provider record added to the DB?

Reply:
Yes there might be a new provider or any update on the already processed provider